File: PolygonProject/V2/Polygon.java
/**
 * <h1>Polygon.java - Assignment 2</h1>
 * The Polygon object extension of PlanarShape
 * Modified: 17-APR-2018
 *
 * @author Cody Lewis - c3283349
 * @version 1.0
 * @since 02-APR-2018
 */
public class Polygon extends PlanarShape{
  /**
   * Default Constructor
   */
  public Polygon(){
    super();
  }
  /**
   * Input Constructor
   * @param p the array of points to be in this object
   */
  public Polygon(Point p[]){
    super();
    for(Point cp : p){
      addPoint(cp);
    }
  }
  /**
   * Get a String representing the Polygon
   * @return A String representing the Polygon
   */
  @Override
  public String toString(){
    String str = "POLY=[";
    if(!points.empty()){
      Queue<Point> cp = new Queue<Point>();
      while(!points.empty()){
        str = str + points.peek().toString();
        cp.enqueue(points.dequeue());
      }
      points = cp;
    }
    str = str + "]:" + String.format("%5.2f",area());
    return str;
  }
  /**
   * Calculate the area of the Polygon
   * @return the area of the Polygon as a double
   */
  @Override
  public double area(){
    if(points.empty()){
      return 0;
    } else {
      Queue<Point> cp = new Queue<Point>();
		  Point i,nextI;
			double A = 0;
      while(!points.empty()){
      	// (1/2)*ABS(SIGMA(i=0)(n-2)(x_i+1 + x_i)*(y_i+1 - y_i))
				i = points.dequeue();
        nextI = (points.empty() && !cp.empty()) ? cp.peek() : points.peek();
        if(nextI == null){ // case of single point
          return 0;
        }
        A += (nextI.getX() + i.getX())*(nextI.getY() - i.getY());
        cp.enqueue(i);
      }
      points = cp;
      A = 0.5*Math.abs(A);
			return A;
    }
  }
  /**
   * Find how far from the origin this Polygon is
   * @return the distance from the origin as a double
   */
  @Override
  public double originDistance(){
    if(points.empty()){
      return 0;
    } else {
      Queue<Point> cp = new Queue<Point>();
      double smallVal = points.peek().distFromOrigin();
      cp.enqueue(points.dequeue());
      double runVal;
      while(!points.empty()){
        runVal = points.peek().distFromOrigin();
        if(runVal < smallVal){
          smallVal = runVal;
        }
        cp.enqueue(points.dequeue());
      }
      points = cp;
      return smallVal;
    }
  }
}

File: PolygonProject/V1/Polygon.cpp
/* Polygon.cpp - Assignment 1
 * Author: Cody Lewis - c3283349
 * Date: 10-MAR-2018
 * Last Modified: 24-MAR-2018
 * Description:
 * The implementation of the Polygon object
 */
#include <iostream>
#include <cmath>
#include <string>
#include <sstream>
#include "ComparePoly.h"
#include "Polygon.h"
#include "Point.h"
#include "Queue.h"
namespace lewis_a1 {
  // Constructors
  Polygon::Polygon(){
    points = new Queue<Point>(); 
  }
  Polygon::Polygon(const Queue<Point>& q){
    points = new Queue<Point>(q);
  }
  Polygon::Polygon(Polygon& p){
    points = new Queue<Point>(*p.getPoints());
  }
  // Destructors
  Polygon::~Polygon(){
    delete points;
  }
  // Mutators
  void Polygon::addPoint(Point& p){
    points->push(p);   
  }
  void Polygon::addPoint(Queue<Point>*& q){
    *points+=*q; // somehow copy to here
  }
  bool Polygon::addPoint(Polygon& p){
    Queue<Point>* q = p.getPoints();
    addPoint(q);
    return true;
  }
  // Queries
  double Polygon::distFromOrigin() const{
    double smallDist,a,b,c;
    bool firstPass = true;
    Queue<Point>* pointsCp = new Queue<Point>(*points);
    while(!pointsCp->isEmpty()){
      a = pointsCp->front().getX();
      b = pointsCp->front().getY();
      c = a*a + b*b;
      c = sqrt(c);
      if(firstPass){
        smallDist = c;
        firstPass = false;
      } else {
        if(c < smallDist){
          smallDist = c;
        }
      }
      pointsCp->pop();
    }
    delete pointsCp;
    return smallDist;
  }
  double Polygon::findArea() const{
  // A = (1/2)*ABS(Sigma(i=0,n-2)((x_i + x_i+1)*(y_i - y_i+1)))
    if(points->isEmpty()){
      return 0;
    } else {
      double area,summand;
      area = 0;
      Queue<Point>* pointsCp = new Queue<Point>(*points);
      Point curr,next,cp;
      cp = pointsCp->front();
      for(curr = pointsCp->front(); !pointsCp->isEmpty(); curr = pointsCp->front()){
        pointsCp->pop();
        if(pointsCp->isEmpty()){
          next = cp; // repeat the first point
        } else {
          next = pointsCp->front();
        }
        summand = (curr.getX() + next.getX()) * (curr.getY() - next.getY());
        if(summand < 0){ 
          summand*=-1; // Absolute value
        }
        area+=summand;
      }
      area*=0.5;
      delete pointsCp;
      return area;
    }
  }
  std::string Polygon::to52F(std::string fPoint) const{ // private
    if(fPoint.find(".") < fPoint.length()){
      if(fPoint.length() - fPoint.find(".") > 3){ // for .2f
        fPoint.erase(fPoint.find(".")+3);
      }
    }
    while(fPoint.length() < 5){ // for 5f
      fPoint = " " + fPoint;
    }
    return fPoint;  
  }
  bool Polygon::isEmpty() const{
    return points->isEmpty();
  }
  Queue<Point>*& Polygon::getPoints(){
    return points;
  }
  std::string Polygon::toString() const{
    std::stringstream ss;
    ss << "[";
    Queue<Point>* pointsCp = new Queue<Point>(*points);
    while(!pointsCp->isEmpty()){
      ss << pointsCp->front().toString() << ",";
      pointsCp->pop();
    }
    delete pointsCp;
    std::string str = ss.str();
    if(!points->isEmpty()){
      str.erase(str.length()-1); // remove trailing comma
    }
    str = str + "]:";
    ss.str(std::string()); // reset ss
    ss << findArea();
    std::string area = ss.str();
    area = to52F(area); // convert area to 5.2f
    str = str + area;
    return str;
  }
  // operator overloads
  bool operator>(const Polygon& lhs,const Polygon& rhs){
    double A = lhs.findArea();
    double minima = A*0.9995; // range is + or - 0.05% of the Area
    double maxima = A*1.0005;
    double A2 = rhs.findArea(); // stops redundant calculations
    if(A2 >= minima && A2 <= maxima){
      return (lhs.distFromOrigin() < rhs.distFromOrigin()); // true if this is closest to the origin
    } else {
      return (A > A2);
    }
  }
  bool operator<(const Polygon& lhs,const Polygon& rhs){
    return !(lhs>rhs);
  }
  std::istream& operator>>(std::istream& in,Polygon& p){
    Queue<Point>* newQ = new Queue<Point>();
    in >> *newQ;
    p.addPoint(newQ);
    delete newQ;
    return in;
  }
}
